### ISSUE TITLE
Collection::update inconsistencies between OnHeapPersistence and DiskPersistence

### ISSUE THREAD

**Issue by codingchili** [npgall/cqengine#158]
Hello!

I'm having issues updating an object in the collection. The update method does not seem to be removing existing items when using DiskPersistence.

Some notes
- The old object is removed successfully if changed to OnHeapPersistence
- The old object is removed successfully if they both match in serialized form
  - If the price of the second car is changed to match the first it works as expected.

It seems the primary key/hashCode/equals is not used for updates when using DiskPersistence?

I created a small reproducer here
https://github.com/codingchili/cqengine/commit/227e07b8217b8e6b5490358e0c9c923163a92553

Is this a bug? Am I doing something wrong?

Thanks

**Comment by npgall** [comment]
This is not a bug as far as I can see, but it's an interesting question! Thanks very much for providing a test case, which made it very easy to investigate!

First of all, I think you might be using the `update()` method incorrectly (you can find javadocs [here](http://htmlpreview.github.io/?http://raw.githubusercontent.com/npgall/cqengine/master/documentation/javadoc/apidocs/com/googlecode/cqengine/package-summary.html)). However you're also encountering some subtle intricacies which relate to object equality and your choice of primary key, which I'll explain below.

The intended usage is: `update(objectsToRemove, objectsToAdd)`

So the following code...
```java
update(singleton(first), singleton(first));
update(singleton(second), singleton(second));
```
...is unusual usage, because it attempts to replace the first object with itself, and then replace the second object with itself. Since the collection will initially not contain either of these objects, the expected outcome is that both objects will be added, and the collection will then contain two objects.

**Plot twist**
A plot twist however, is that in your example, you also initialize both the first and second `Car` objects with the same `carId`. The `Car` object implements the `equals()` method to determine equality based only on the value of `carId`. Also, the primary key with which you configure the `DiskPersistence` is based on `carId` as well.

If you really want to store different `Car` objects in the collection with the same `carId` but different values in other fields, then:

1. You need to implement the `equals()` method to compare the other fields as well so it will be consistent with your application's notion of equality, and
2. If you use `DispPersistence` then you will not be able to use `carId` as a primary key at all, because it will not - but should - uniquely identify objects.

So you might ask then, why does the use of `DiskPersistence` result in two objects in the collection in this example? Well as mentioned above, that's actually expected for this usage of `update()` according to its documentation. However you are encountering an internal implementation detail of CQEngine wherein the unique constraint applied by `DiskPersistence` is actually a combination of (primary key, serialized form). This means if `DiskPersistence` is configured with an invalid primary key which does not uniquely identify objects, then technically two non-identical objects can be added to the collection with the same primary key.

A second question is why does the collection only contain one object and not two, when `OnHeapPersistence` is used? That is because the implementation of the `equals()` method is not consistent with the application's notion of equality. If you modify `equals()` to consider the other fields when determining equality, then collection will contain both objects when the `OnHeapPersistence` is used.

**TL;DR**
If you ensure that your implementation of the `equals()` method in your domain objects, and your choice of primary key, will both uniquely identify objects, then this issue should disappear. Let me know if you still encounter issues though.

Hope that helps!

Best regards,
Niall

**Comment by codingchili** [comment]
Hello! Thanks alot for the quick and in-depth reply.

I did not expect the combination of (primary key, serialized form) for DiskPersistence. For my use case I use CQEngine as a map, where I only care about the primary key when checking for equality (hashCode/equals). The update method is being used to add if not present, or overwrite any existing item in the collection based only on its primary key.

To make this work with DiskPersistence, I would retrieve the old object based on the primary key of the new object, and then remove the old object before adding the new object. Is there any alternative to this? I liked the update method because I'm under the impression that its more efficient and handles concurrency better.

[example Map::put operation using CQ collection](https://github.com/codingchili/chili-core/blob/master/core/main/java/com/codingchili/core/storage/IndexedMap.java#L88)

**Comment by npgall** [comment]
The TransactionalIndexedCollection does offer support to replace objects atomically only if the old object is already contained in the collection. Basically, the update() method in that implementation supports an optional STRICT_REPLACEMENT parameter to request that behaviour:

http://htmlpreview.github.io/?http://raw.githubusercontent.com/npgall/cqengine/master/documentation/javadoc/apidocs/com/googlecode/cqengine/TransactionalIndexedCollection.html#update-java.lang.Iterable-java.lang.Iterable-com.googlecode.cqengine.query.option.QueryOptions-

However unfortunately this is not natively supported by the ConcurrentIndexedCollection, because there's no way to implement atomic putIfAbsent behaviour across multiple data structures (the collection itself plus various indexes), without introducing some kind of coordination between threads (like transactions). 

(Effectively, this is precisely what the TransactionalIndexedCollection does. It supports batch operations on the collection which appear atomic to concurrent reading threads using an MVCC algorithm, at the expense of limited concurrency for writing threads.)

So if you need atomic replacement the update() method is the way to go, but you might need to switch to a different implementation of the IndexedCollection. You could also consider the ObjectLockingIndexedCollection. You can read about the differences in: 

http://htmlpreview.github.io/?http://raw.githubusercontent.com/npgall/cqengine/master/documentation/javadoc/apidocs/com/googlecode/cqengine/package-summary.html
